[Filebeat] Enable non-AWS S3 buckets for aws-s3 input

legoguy1000 opened the pull request:
## What does this PR do?

This PR adds the ability to use non-AWS S3 bucket services with the aws-s3 input. With the new polling feature the input can now poll other S3 providers like Minio that use both virtual host style and path style buckets as well as the ability to use non SSL enabled buckets (if wanted).  This also updates the `log.file.path` field to use the actual URL used to download the file instead of manually generating it with hard coded values like `amazonaws.com`.

## Why is it important?

To be able to ingest data from other S3 providers besides AWS.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## How to test this PR locally


## Related issues

- Closes #28222 

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-28234/detail/PR-28234/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-28234/detail/PR-28234/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-28234/detail/PR-28234/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-28234/detail/PR-28234/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2021-10-04T14:02:14.034Z&rangeTo=2021-10-04T14:22:14.034Z&transactionName=Beats/beats/PR-{number}&transactionType=unknown&latencyAggregationType=avg&traceId=9a344aa42ac7ecd95ab9dda12723e4de&transactionId=82b4874466175b79)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats



* Reason: `The PR is not allowed to run in the CI yet`


* Start Time: 2021-10-04T14:12:14.034+0000


* Duration: 4 min 18 sec


* Commit: f4a13fd90a705c316f7aa2cfddaa3a0a631cf7d5




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












## :robot: GitHub comments

To re-run your PR in the CI, just comment with:


- `/test` : Re-trigger the build.

- `/package` : Generate the packages and run the E2E tests.

- `/beats-tester` : Run the installation tests with beats-tester.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

aspacca commented:
@legoguy1000 thanks for the contribution, the feature is welcome.

I think we can anyway simplify the implementation and changes required with no need to introduce a `non_aws_bucket_name` that can be confusing.

My suggestion is to let the `endpoint` accept a full URI and then parse it and check if it contains a scheme.
In this case we will just use it without using the endpoint enrichment.

With pseudo-code something along this:
```
endpoint := awscommon.EnrichAWSConfigWithEndpoint(in.config.AWSConfig.Endpoint, s3ServiceName, in.awsConfig.Region, in.awsConfig)
if url.Parse(in.config.AWSConfig.Endpoint).Scheme != "" {
    endpoint = awssdk.ResolveWithEndpointURL(in.config.AWSConfig.Endpoint)
}
```

`path_style` support is it fine to be added.

legoguy1000 commented:
If not add something like `non_aws_bucket_name` would u just use the `bucket_arn` to set the bucket name? My biggest reason to add a new field was to be able to easily identify that we don't want to use AWS specific settings and options.

legoguy1000 commented:
@aspacca i removed the `non_aws_bucket_name` config and added validation via the enpoint field.  I also added some config overrides for the region and provider to make sure the `cloud.*` are relatively accurate and not misleading/hardcoded.

aspacca commented:
@legoguy1000 

>If not add something like non_aws_bucket_name would u just use the bucket_arn to set the bucket name? 

you are right about this: we had a discussion internally and between introducing a breaking change renaming `bucket_arn` to `bucket_name` and use for both cases we decided it is actually better your initial solution using `non_aws_bucket_name`

I apologise for asking you to revert your change for it: you can add back `non_aws_bucket_name`, thanks

aspacca commented on x-pack/filebeat/input/awss3/config.go:
```diff
@@ -33,6 +34,9 @@ type config struct {
 	FileSelectors       []fileSelectorConfig `config:"file_selectors"`
 	ReaderConfig        readerConfig         `config:",inline"` // Reader options to apply when no file_selectors are used.
+	PathStyle           bool                 `config:"path_style"`
+	RegionOverride      string               `config:"region_override"`
```
I would assume S3 compatible buckets support https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketLocation.html

Could you help me validate this?
`aws s3api --endpoint=%ENDPOINT% get-bucket-location --bucket %BUCKET_NAME%`
Does it return the following?
```
{
    "LocationConstraint": "%REGION%"
}
```

If so we should not need `RegionOverride` but just rely on `getRegionForBucketARN` also for not AWS S3 buckets

legoguy1000 replied:
Possibly.  I know minio has a region setting. I'll change it and see what happens.

legoguy1000 replied:
yes minio does and looks like all the s3 services i found also do.

aspacca commented on x-pack/filebeat/input/awss3/config.go:
```diff
@@ -33,6 +34,9 @@ type config struct {
 	ReaderConfig        readerConfig         `config:",inline"` // Reader options to apply when no file_selectors are used.
+	PathStyle           bool                 `config:"path_style"`
+	RegionOverride      string               `config:"region_override"`
+	ProviderOverride    string               `config:"provider_override"`
```
Is there anyway we can infer the provider from the endpoint or any API request?
If we can avoid to specify it explicitly and let the code find it it would be better, what do you think?

legoguy1000 replied:
I think it would be hard unless you just Didi it by domain name but even then it would make `aws` into `amazonaws`

aspacca commented on x-pack/filebeat/input/awss3/input.go:
```diff
@@ -199,15 +199,25 @@ func (in *s3Input) createS3Lister(ctx v2.Context, cancelCtx context.Context, cli
-	if err != nil {
-		return nil, fmt.Errorf("failed to get AWS region for bucket_arn: %w", err)
+	var s3Client *s3.Client
+	endpoint, _ := url.Parse(in.config.AWSConfig.Endpoint)
```
I would move this logic to `awscommon.EnrichAWSConfigWithEndpoint` returning directly either the composed endpoint or the raw one if it's a full URI

legoguy1000 commented:
@aspacca let me know what u think of the changes. Instead of `non_aws_bucket_name` I just added back `bucket_name`. It will then work with AWS or non AWS buckets especially since the ARN value isn't really used except to pull the bucket name from it.
